modules/xcm-bridge-hub-router/src/lib.rs

```diff
@@ -0,0 +1,429 @@
+	/// ```
+	///
+	/// If at least one of other channels is congested, the local XCM channel with sibling
+	/// bridge hub eventually becomes congested too. And we have no means to detect - which
+	/// bridge exactly causes the congestion. So the best solution here is not to make
+	/// any differences between all bridges, started by this chain.
+	#[pallet::storage]
+	#[pallet::getter(fn delivery_fee_factor)]
+	pub type DeliveryFeeFactor<T: Config<I>, I: 'static = ()> =
+		StorageValue<_, FixedU128, ValueQuery, InitialFactor>;
+
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		/// Called when new message is sent (queued to local outbound XCM queue) over the bridge.
+		pub(crate) fn on_message_sent_to_bridge(message_size: u32) {
+			// if outbound queue is not congested, do nothing
+			if !T::WithBridgeHubChannel::is_congested() {
+				return
+			}
+
+			// ok - we need to increase the fee factor, let's do that
+			let message_size_factor = FixedU128::from_u32(message_size.saturating_div(1024))
+				.saturating_mul(MESSAGE_SIZE_FEE_BASE);
+			let total_factor = EXPONENTIAL_FEE_BASE.saturating_add(message_size_factor);
+			DeliveryFeeFactor::<T, I>::mutate(|f| {
+				let previous_factor = *f;
+				*f = f.saturating_mul(total_factor);
+				log::info!(
+					target: LOG_TARGET,
+					"Bridge queue is congested. Increased fee factor from {} to {}",
+					previous_factor,
+					f,
+				);
+				*f
+			});
+		}
+	}
+}
+
+/// We'll be using `SovereignPaidRemoteExporter` to send remote messages over the sibling/child
+/// bridge hub.
+type ViaBridgeHubExporter<T, I> = SovereignPaidRemoteExporter<
+	Pallet<T, I>,
+	<T as Config<I>>::ToBridgeHubSender,
+	<T as Config<I>>::UniversalLocation,
+>;
+
+// This pallet acts as the `ExporterFor` for the `SovereignPaidRemoteExporter` to compute
+// message fee using fee factor.
+impl<T: Config<I>, I: 'static> ExporterFor for Pallet<T, I> {
+	fn exporter_for(
+		network: &NetworkId,
+		_remote_location: &InteriorMultiLocation,
+		message: &Xcm<()>,
+	) -> Option<(MultiLocation, Option<MultiAsset>)> {
+		// ensure that the message is sent to the expected bridged network
+		if *network != T::BridgedNetworkId::get() {
+			return None
+		}
+
+		// compute fee amount. Keep in mind that this is only the bridge fee. The fee for sending
+		// message from this chain to child/sibling bridge hub is determined by the
+		// `Config::ToBridgeHubSender`
+		let message_size = message.encoded_size();
+		let message_fee = (message_size as u128).saturating_mul(T::ByteFee::get());
+		let fee_sum = T::BaseFee::get().saturating_add(message_fee);
+		let fee_factor = Self::delivery_fee_factor();
+		let fee = fee_factor.saturating_mul_int(fee_sum);
+
+		log::info!(
+			target: LOG_TARGET,
+			"Going to send message ({} bytes) over bridge. Computed bridge fee {} using fee factor {}",
+			fee,
+			fee_factor,
+			message_size,
+		);
+
+		Some((T::SiblingBridgeHubLocation::get(), Some((T::FeeAsset::get(), fee).into())))
```

Comment:
@svyatonik 

Sorry, late question. I was looking on this while working on https://github.com/paritytech/parity-bridges-common/issues/2554.

If I understand correctly, the fee that we set here will be used through the `ViaBridgeHubExporter` -> `SovereignPaidRemoteExporter` [here](https://github.com/paritytech/polkadot-sdk/blob/95052437800dec7934f6804a77053e4552fc28e5/polkadot/xcm/xcm-builder/src/universal_exports.rs#L305-L309):

```
				WithdrawAsset(fees.clone().into()),
				BuyExecution { fees, weight_limit: Unlimited },
				export_instruction,
				RefundSurplus,
				DepositAsset { assets: All.into(), beneficiary: local_from_bridge },
``` 

But what will happen to it exactly ? `BuyExecution`, doesn't seem to do anything with it ([here](https://github.com/paritytech/polkadot-sdk/blob/95052437800dec7934f6804a77053e4552fc28e5/polkadot/xcm/xcm-executor/src/lib.rs#L687-L690)) so I suppose that it will be deposited to the sovereign account of the asset hub on the bridge hub when `DepositAsset { assets: All.into(), beneficiary: local_from_bridge }` is executed ?

Did I understand correctly ? And is this the desired behavior ?

Reply:
@serban300 IIRC `Unlimited` is changed to **actual** weight of the XCM program by [`AllowTopLevelPaidExecutionFrom` barrier](https://github.com/paritytech/polkadot-sdk/blob/676bacd7299826fb0ea88055d2108f3c6a734291/polkadot/xcm/xcm-builder/src/barriers.rs#L98-L101). So idea behind that is following:
1) here, at asset hub, user pays fee that covers both XCM execution at sibling BH (the weight we're talking here) + some extra;
2) when message is dispatched at the sibling BH, xcm executor [computes the actual weight of the message](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-executor/src/lib.rs#L185-L192) and ["sends" it to the `AllowTopLevelPaidExecutionFrom`](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-executor/src/lib.rs#L208-L213);
3) then the `AllowTopLevelPaidExecutionFrom` changes `Unlimited` to actual weight, so we'll be charging the SA of AH at BH in the end. But without that extra that was paid by the user at the AH.

Please correct me here, if you see any mistake - I was checking it long time ago, so could lie (accidentally) here :)

Reply:
Oh you're right. Sorry, I forgot about `AllowTopLevelPaidExecutionFrom`. So the weight of the message is indeed charged. 

But what about the extra amount ? Will it be deposited to the SA of AH at BH ? Especially if the bridge queue is congested and we apply a high fee factor at the AH, the extra amount can be significant.